### ISSUE TITLE
[Bug]: Inference results on CPU and GPU are different

### ISSUE THREAD

**Issue by nakayamarusu** [openvinotoolkit/openvino#23523]
### OpenVINO Version

2024.0.0

### Operating System

Windows System

### Device used for inference

CPU

### Framework

None

### Model used

Anomalib PaDiM

### Issue description

# **Problem**
Learn images of MVTec bottles using Padim. And after exporting it to a file that can be used in Openvino, I ran the inference in Openvino. You can get the correct heatmap by inferring on the CPU, but it won't work on the GPU. I just changed the device used from CPU to GPU. The GPU uses Intel iRIS Xe, and we have confirmed that OpenVino is compatible.

### Step-by-step reproduction

# **Operating environment**
- Windows11 Pro
- Python 3.10.13
- Anomalib 1.0.0
- OpenVINO 2024.0.0
- torch 2.1.2
- torchmetrics 0.10.3
- torchvision 0.16.2
- pytorch-lightning 2.2.1
- onnx 1.15.0

# **Train**
Learning was performed using the following command
`anomalib fit -c configs/model/padim.yaml --data configs/folder_bottle.yaml`

▼ padim.yaml
```
model:
  class_path: anomalib.models.Padim
  init_args:
    layers:
      - layer1
      - layer2
      - layer3
    #input_size: [256, 256]
    backbone: resnet18
    pre_trained: true
    n_features: null

metrics:
  pixel: AUROC
```

▼ folder_bottle_yaml
```
class_path: anomalib.data.Folder
init_args:
  name: bottle
  root: "dataset/bottle"
  normal_dir: "train/good"
  abnormal_dir: "test/broken_large"
  normal_test_dir: "test/good"
  mask_dir: "ground_truth/broken_large"
  normal_split_ratio: 0
  extensions: [".png"]
  image_size: [256, 256]
  #center_crop: null
  #normalization: imagenet
  train_batch_size: 32
  eval_batch_size: 32
  num_workers: 8
  task: CLASSIFICATION
  #transform_config_train: null
  #transform_config_eval: null
  test_split_mode: NONE
  test_split_ratio: 0.2
  val_split_mode: same_as_test
  val_split_ratio: 0.5
  seed: null
```

# **Export**
Converted to a model that can be used with OpenVINO using the following command

`anomalib export --model Padim --export_type OPENVINO --ckpt_path　results/Padim/bottle/latest/weights/lightning/model.ckpt`

# **Inference**
The inference was made by creating python code as follows.
```python
from anomalib.deploy.inferencers import OpenVINOInferencer
import cv2
import numpy as np
import os

inferencer = OpenVINOInferencer(
    path=r"C:\anomalib_v1\results\weights\openvino\model.onnx",
    metadata=r"C:\anomalib_v1\results\weights\openvino\metadata.json",
    device="GPU",
)

img_path = r"C:\anomalib_v1\dataset\bottle\test\broken_small"

for i, file in enumerate(os.listdir(img_path)):
    
    image = cv2.imread(img_path + "\\" + file)
    image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
    input_img = image.astype(np.float32) / 255.0 
    result = inferencer.predict(image = input_img)
    print(i, result.pred_score)
    cv2.imshow("heatmap", cv2.cvtColor(result.heat_map, cv2.COLOR_RGB2BGR))
    cv2.waitKey()
```

# **Inference result**
Heat map when the device used for inference is CPU as shown below

```python
inferencer = OpenVINOInferencer(
    path=r"C:\anomalib_v1\results\weights\openvino\model.onnx",
    metadata=r"C:\anomalib_v1\results\weights\openvino\metadata.json",
    device="CPU",
)
```
![heatmap1](https://github.com/openvinotoolkit/anomalib/assets/163484775/85e720ff-924b-46ef-afe9-82b79bca0913)
▲ pred_score : 0.4836

![heatmap2](https://github.com/openvinotoolkit/anomalib/assets/163484775/4f5ba28b-cb75-4d65-b5e2-5c17080091b6)
▲ pred_score : 0.5605

Heat map when the device used for inference is GPU as shown below

```python
inferencer = OpenVINOInferencer(
    path=r"C:\anomalib_v1\results\weights\openvino\model.onnx",
    metadata=r"C:\anomalib_v1\results\weights\openvino\metadata.json",
    device="GPU",
)
```
![heatmap3](https://github.com/openvinotoolkit/anomalib/assets/163484775/abb0ace8-3cde-4f01-8a68-e3daf234fb3d)
▲ pred_score : 0.0

![heatmap4](https://github.com/openvinotoolkit/anomalib/assets/163484775/29b6c949-a7fc-4f6d-8b0a-5b368d7e8907)
▲ pred_score : 0.0

### Relevant log output

```shell
(anomalib_latest_env) C:\anomalib_v1>anomalib fit -c configs/model/padim.yaml --data configs/data/folder_bottle.yaml
2024-03-15 10:37:21,750 - anomalib.utils.config - WARNING - Anomalib currently does not support multi-gpu training. Setting devices to 1.
[03/15/24 10:37:21] WARNING  Anomalib currently does not support multi-gpu training. Setting devices to 1.     config.py:126
C:\Users\n-nakayama\AppData\Local\anaconda3\envs\anomalib_latest_env\lib\site-packages\torchmetrics\utilities\prints.py:36: UserWarning: Metric `PrecisionRecallCurve` will save all targets and predictions in buffer. For large datasets this may lead to large memory footprint.
  warnings.warn(*args, **kwargs)
2024-03-15 10:37:21,781 - anomalib.models.components.base.anomaly_module - INFO - Initializing Padim model.
                    INFO     Initializing Padim model.                                                  anomaly_module.py:39
2024-03-15 10:37:21,954 - timm.models.helpers - INFO - Loading pretrained weights from url (https://download.pytorch.org/models/resnet18-5c106cde.pth)
                    INFO     Loading pretrained weights from url                                              helpers.py:247
                             (https://download.pytorch.org/models/resnet18-5c106cde.pth)
2024-03-15 10:37:22,064 - anomalib.callbacks - INFO - Loading the callbacks
[03/15/24 10:37:22] INFO     Loading the callbacks                                                            __init__.py:43
2024-03-15 10:37:22,064 - anomalib.engine.engine - INFO - Overriding max_epochs from None with 1 for Padim
                    INFO     Overriding max_epochs from None with 1 for Padim                                   engine.py:84
2024-03-15 10:37:22,082 - anomalib.engine.engine - INFO - Overriding val_check_interval from None with 1.0 for Padim
                    INFO     Overriding val_check_interval from None with 1.0 for Padim                         engine.py:84
2024-03-15 10:37:22,084 - anomalib.engine.engine - INFO - Overriding num_sanity_val_steps from None with 0 for Padim
                    INFO     Overriding num_sanity_val_steps from None with 0 for Padim                         engine.py:84
2024-03-15 10:37:22,175 - lightning.pytorch.utilities.rank_zero - INFO - GPU available: False, used: False
[03/15/24 10:37:22] INFO     GPU available: False, used: False                                               rank_zero.py:64
2024-03-15 10:37:22,187 - lightning.pytorch.utilities.rank_zero - INFO - TPU available: False, using: 0 TPU cores
                    INFO     TPU available: False, using: 0 TPU cores                                        rank_zero.py:64
2024-03-15 10:37:22,191 - lightning.pytorch.utilities.rank_zero - INFO - IPU available: False, using: 0 IPUs
                    INFO     IPU available: False, using: 0 IPUs                                             rank_zero.py:64
2024-03-15 10:37:22,191 - lightning.pytorch.utilities.rank_zero - INFO - HPU available: False, using: 0 HPUs
                    INFO     HPU available: False, using: 0 HPUs                                             rank_zero.py:64
2024-03-15 10:37:22,191 - lightning.pytorch.utilities.rank_zero - INFO - `Trainer(val_check_interval=1.0)` was configured so validation will run at the end of the training epoch..
                    INFO     `Trainer(val_check_interval=1.0)` was configured so validation will run at the  rank_zero.py:64
                             end of the training epoch..
C:\Users\n-nakayama\AppData\Local\anaconda3\envs\anomalib_latest_env\lib\site-packages\torchmetrics\utilities\prints.py:36: UserWarning: Metric `ROC` will save all targets and predictions in buffer. For large datasets this may lead to large memory footprint.
  warnings.warn(*args, **kwargs)
C:\Users\n-nakayama\AppData\Local\anaconda3\envs\anomalib_latest_env\lib\site-packages\lightning\pytorch\core\optimizer.py:180: `LightningModule.configure_optimizers` returned `None`, this fit will run with no optimizer
2024-03-15 10:37:22,375 - lightning.pytorch.callbacks.model_summary - INFO -
  | Name                  | Type                     | Params
-------------------------------------------------------------------
0 | model                 | PadimModel               | 2.8 M
1 | _transform            | Compose                  | 0
2 | normalization_metrics | MinMax                   | 0
3 | image_threshold       | F1AdaptiveThreshold      | 0
4 | pixel_threshold       | F1AdaptiveThreshold      | 0
5 | image_metrics         | AnomalibMetricCollection | 0
6 | pixel_metrics         | AnomalibMetricCollection | 0
-------------------------------------------------------------------
2.8 M     Trainable params
0         Non-trainable params
2.8 M     Total params
11.131    Total estimated model params size (MB)
                    INFO                                                                                 model_summary.py:90
                               | Name                  | Type                     | Params
                             -------------------------------------------------------------------
                             0 | model                 | PadimModel               | 2.8 M
                             1 | _transform            | Compose                  | 0
                             2 | normalization_metrics | MinMax                   | 0
                             3 | image_threshold       | F1AdaptiveThreshold      | 0
                             4 | pixel_threshold       | F1AdaptiveThreshold      | 0
                             5 | image_metrics         | AnomalibMetricCollection | 0
                             6 | pixel_metrics         | AnomalibMetricCollection | 0
                             -------------------------------------------------------------------
                             2.8 M     Trainable params
                             0         Non-trainable params
                             2.8 M     Total params
                             11.131    Total estimated model params size (MB)
C:\Users\n-nakayama\AppData\Local\anaconda3\envs\anomalib_latest_env\lib\site-packages\lightning\pytorch\trainer\connectors\data_connector.py:436: Consider setting `persistent_workers=True` in 'train_dataloader' to speed up the dataloader worker initialization.
C:\Users\n-nakayama\AppData\Local\anaconda3\envs\anomalib_latest_env\lib\site-packages\lightning\pytorch\trainer\connectors\data_connector.py:436: Consider setting `persistent_workers=True` in 'val_dataloader' to speed up the dataloader worker initialization.
Epoch 0:   0%|                                                                                       | 0/7 [00:00<?, ?it/s]C:\Users\n-nakayama\AppData\Local\anaconda3\envs\anomalib_latest_env\lib\site-packages\lightning\pytorch\loops\optimization\automatic.py:129: `training_step` returned `None`. If this was on purpose, ignore this warning...
Epoch 0: 100%|███████████████████████████████████████████████████████████████████████████████| 7/7 [00:17<00:00,  0.39it/s]2024-03-15 10:37:42,083 - anomalib.models.image.padim.lightning_model - INFO - Aggregating the embedding extracted from the training set.
[03/15/24 10:37:42] INFO     Aggregating the embedding extracted from the training set.                lightning_model.py:87
2024-03-15 10:37:42,109 - anomalib.models.image.padim.lightning_model - INFO - Fitting a Gaussian to the embedding collected from the training set.
                    INFO     Fitting a Gaussian to the embedding collected from the training set.      lightning_model.py:90
                                                                                                                           C:\Users\n-nakayama\AppData\Local\anaconda3\envs\anomalib_latest_env\lib\site-packages\torchmetrics\utilities\prints.py:36: UserWarning: No negative samples in targets, false positive value should be meaningless. Returning zero tensor in false positive score
  warnings.warn(*args, **kwargs)
Epoch 0: 100%|████████████████████████████████████████████████████████████| 7/7 [00:37<00:00,  0.19it/s, pixel_AUROC=0.976]2024-03-15 10:38:00,833 - lightning.pytorch.utilities.rank_zero - INFO - `Trainer.fit` stopped: `max_epochs=1` reached.
[03/15/24 10:38:00] INFO     `Trainer.fit` stopped: `max_epochs=1` reached.                                  rank_zero.py:64
Epoch 0: 100%|████████████████████████████████████████████████████████████| 7/7 [00:37<00:00,  0.19it/s, pixel_AUROC=0.976]
2024-03-15 10:38:03,269 - anomalib.callbacks.timer - INFO - Training took 40.91 seconds
[03/15/24 10:38:03] INFO     Training took 40.91 seconds                                                         timer.py:59

(anomalib_latest_env) C:\anomalib_v1>anomalib export --model Padim --export_type OPENVINO --ckpt_path results/Padim/bottle/latest/weights/lightning/model.ckpt
C:\Users\n-nakayama\AppData\Local\anaconda3\envs\anomalib_latest_env\lib\site-packages\torchmetrics\utilities\prints.py:36: UserWarning: Metric `PrecisionRecallCurve` will save all targets and predictions in buffer. For large datasets this may lead to large memory footprint.
  warnings.warn(*args, **kwargs)
2024-03-15 10:40:58,023 - anomalib.models.components.base.anomaly_module - INFO - Initializing Padim model.
[03/15/24 10:40:58] INFO     Initializing Padim model.                                                  anomaly_module.py:39
2024-03-15 10:40:58,198 - timm.models.helpers - INFO - Loading pretrained weights from url (https://download.pytorch.org/models/resnet18-5c106cde.pth)
                    INFO     Loading pretrained weights from url                                              helpers.py:247
                             (https://download.pytorch.org/models/resnet18-5c106cde.pth)
2024-03-15 10:40:58,291 - anomalib.callbacks - INFO - Loading the callbacks
                    INFO     Loading the callbacks                                                            __init__.py:43
2024-03-15 10:40:58,291 - anomalib.engine.engine - INFO - Overriding max_epochs from None with 1 for Padim
                    INFO     Overriding max_epochs from None with 1 for Padim                                   engine.py:84
2024-03-15 10:40:58,291 - anomalib.engine.engine - INFO - Overriding val_check_interval from None with 1.0 for Padim
                    INFO     Overriding val_check_interval from None with 1.0 for Padim                         engine.py:84
2024-03-15 10:40:58,291 - anomalib.engine.engine - INFO - Overriding num_sanity_val_steps from None with 0 for Padim
                    INFO     Overriding num_sanity_val_steps from None with 0 for Padim                         engine.py:84
2024-03-15 10:40:58,400 - lightning.pytorch.utilities.rank_zero - INFO - GPU available: False, used: False
[03/15/24 10:40:58] INFO     GPU available: False, used: False                                               rank_zero.py:64
2024-03-15 10:40:58,400 - lightning.pytorch.utilities.rank_zero - INFO - TPU available: False, using: 0 TPU cores
                    INFO     TPU available: False, using: 0 TPU cores                                        rank_zero.py:64
2024-03-15 10:40:58,400 - lightning.pytorch.utilities.rank_zero - INFO - IPU available: False, using: 0 IPUs
                    INFO     IPU available: False, using: 0 IPUs                                             rank_zero.py:64
2024-03-15 10:40:58,400 - lightning.pytorch.utilities.rank_zero - INFO - HPU available: False, using: 0 HPUs
                    INFO     HPU available: False, using: 0 HPUs                                             rank_zero.py:64
2024-03-15 10:40:58,400 - lightning.pytorch.utilities.rank_zero - INFO - `Trainer(val_check_interval=1.0)` was configured so validation will run at the end of the training epoch..
                    INFO     `Trainer(val_check_interval=1.0)` was configured so validation will run at the  rank_zero.py:64
                             end of the training epoch..
2024-03-15 10:40:58,870 - anomalib.models.components.base.anomaly_module - INFO - Initializing Padim model.
                    INFO     Initializing Padim model.                                                  anomaly_module.py:39
2024-03-15 10:40:59,028 - timm.models.helpers - INFO - Loading pretrained weights from url (https://download.pytorch.org/models/resnet18-5c106cde.pth)
[03/15/24 10:40:59] INFO     Loading pretrained weights from url                                              helpers.py:247
                             (https://download.pytorch.org/models/resnet18-5c106cde.pth)
C:\Users\n-nakayama\AppData\Local\anaconda3\envs\anomalib_latest_env\lib\site-packages\torch\onnx\_internal\jit_utils.py:307: UserWarning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied. (Triggered internally at ..\torch\csrc\jit\passes\onnx\constant_fold.cpp:181.)
  _C._jit_pass_onnx_node_shape_type_inference(node, params_dict, opset_version)
C:\Users\n-nakayama\AppData\Local\anaconda3\envs\anomalib_latest_env\lib\site-packages\torch\onnx\utils.py:702: UserWarning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied. (Triggered internally at ..\torch\csrc\jit\passes\onnx\constant_fold.cpp:181.)
  _C._jit_pass_onnx_graph_shape_type_inference(
C:\Users\n-nakayama\AppData\Local\anaconda3\envs\anomalib_latest_env\lib\site-packages\torch\onnx\utils.py:1209: UserWarning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied. (Triggered internally at ..\torch\csrc\jit\passes\onnx\constant_fold.cpp:181.)
  _C._jit_pass_onnx_graph_shape_type_inference(
2024-03-15 10:41:03,818 - root - INFO - Exported model to results\weights\openvino\model.xml
[03/15/24 10:41:03] INFO     Exported model to results\weights\openvino\model.xml                              engine.py:918
```

### Issue submission checklist

- [X] I'm reporting an issue. It's not a question.
- [X] I checked the problem with the documentation, FAQ, open issues, Stack Overflow, etc., and have not found a solution.
- [X] There is reproducer code and related data files such as images, videos, models, etc.

**Comment by wenjiew** [comment]
@MarkByun Could someone from anomalib side take a look? Thanks!

**Comment by nakayamarusu** [comment]
@MarkByun Are you facing the same problem?

**Comment by wenjiew** [comment]
Further follow-up required.

**Comment by ashwinvaidya17** [comment]
@nakayamarusu can you open this on Anomalib's issue page (https://github.com/openvinotoolkit/anomalib/issues) so we can continue there.

**Comment by nakayamarusu** [comment]
> @nakayamarusu can you open this on Anomalib's issue page (https://github.com/openvinotoolkit/anomalib/issues) so we can continue there.

@ashwinvaidya17 Does that mean I should repost to that URL?

**Comment by ashwinvaidya17** [comment]
@nakayamarusu never mind we can just continue here. It does not seem to be related to Anomalib. I don't have an Intel GPU so I can't test this on GPU but I did try this on my `i9-10980XE`.
It seems to be working well.
![image](https://github.com/openvinotoolkit/openvino/assets/17232914/a2db84c0-d2ac-41bf-9ab9-97bac4b2e37d)

One difference is that I tried on the Anomalib's main branch. So not the pip version. Maybe that might help?
